### PR TITLE
Use const ref in order to avoid QgsFeature object copy while evaluating expression

### PR DESCRIPTION
**Pull request by troopa81**
## Description

Return const ref in QgsExpressionContext::getFeature method to avoid QgsFeature object copy and avoid too many call to hasFeature to optimize expression evaluation.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by m-kuhn** [comment]
Do you have evidence that this is actually faster?

Some thoughts:

Copying features is very cheap because implicit sharing is used. A detach of the feature is only happening if a non-const method is called on the feature, which can also be guaranteed by saving the return value in a const feature.

```
const QgsFeature feature = context.feature();
```

Returning a reference is always a bit more dangerous because the caller has to make sure the reference is not living longer than the feature inside the context.

I also wonder if the `feature.isValid()` check is any faster than the `context.hasFeature()` check.

**Comment by nyalldawson** [comment]
> I also wonder if the `feature.isValid()` check is any faster than the `context.hasFeature()` check.

I doubt it -- QgsExpressionContextScope stores a bool indicating whether the scope has a feature set, so it's just returning that.

**Comment by troopa81** [comment]
> Do you have evidence that this is actually faster?

Quite, I benchmark with perf, before and after my modifications on the exact same use case (start QGIS server, trigger exact same GetMap) 

Before
perfs_before
After
perf_after

> Copying features is very cheap because implicit sharing is used

Yes, Indeed. But there is still the copy/destruction of many QgsFeature object itself.

> Returning a reference is always a bit more dangerous because the caller has to make sure the reference is not living longer than the feature inside the context.turning a reference is always a bit more dangerous because the caller has to make sure the reference is not living longer than the feature inside the context.

Sure, but I would say that you get a const reference in the scope of a function most of the time. So, the case you describe is quite rare, no? Or maybe, I miss something

> I also wonder if the feature.isValid() check is any faster than the context.hasFeature() check.

*feature* and *hasFeature* iterates both on contextscope for each feature to test if context has feature. It's not that isValid is faster, it's just that it's enough to call it once instead of 2.

**Comment by m-kuhn** [comment]
That's convincing evidence.

> Sure, but I would say that you get a const reference in the scope of a function most of the time. So, the case you describe is quite rare, no? Or maybe, I miss something

Indeed. It's just a general safety consideration. If it's worth it and if it's an API that can safely be consumed, there are no objections from my side.

File: src/core/qgsexpressioncontext.h [hunk 1]
```diff
@@ -732,6 +732,8 @@ class CORE_EXPORT QgsExpressionContext
 
   private:
 
+    static const QgsFeature missingFeature;
```
Comment (m-kuhn) [inline src/core/qgsexpressioncontext.h]: Can this be moved to a local static to hide this implementation detail from the public headers?
And in either case, please prefix with an `s`.

Comment (troopa81) [reply src/core/qgsexpressioncontext.h]: I removed the static member 8548eab

**Comment by nyalldawson** [comment]
I'd also like to avoid the const ref here if we can, as it seems like a step backwards to me (I can't remember -- does sip make a copy in this situation?).

Looking at the perf charts, it seems like part of the issue is the Q_FOREACH loop used in hasFeature. Does replacing that with a non-copying version help? (e.g. range based for).

How much of a performance boost is there if you leave the rest of your changes, but return a copy of the feature instead of the const ref?

**Comment by troopa81** [comment]
> I'd also like to avoid the const ref here if we can, as it seems like a step backwards to me (I can't remember -- does sip make a copy in this situation?).

I take a look to coding standard and it's well written to return implicit shared object by value and not by reference... my bad.

> How much of a performance boost is there if you leave the rest of your changes, but return a copy of the feature instead of the const ref?

I get this graph with removing the const ref :

output_w_noconstref

So 2.92 instead of 2.26 (versus 7.21 with no modifications), so the main problem is indeed the Q_FOREACH loop. I pushed the modifications ee2cf07

**Comment by nyalldawson** [comment]
The changes to QgsExpressionContextScope will also need rolling back.

What I'm curious about now is whether you see an improvement by leaving the hasFeature check, but changing hasFeature to use a range based loop instead of Q_FOREACH

**Comment by troopa81** [comment]
> What I'm curious about now is whether you see an improvement by leaving the hasFeature check, but changing hasFeature to use a range based loop instead of Q_FOREACH

I try with runnning forward and backward in scopes.


|  |  |
| ----- | ----- |
| no qforeeach forward | 6.98% |
| no qforeeach backward | 3.88% |
| without hasfeature | 2.86% |

I think the results depends a lot of the content of the scope list and the way we run through more than q_foreach or not. If the scope that has feature is in the beginning of the list, it will be quick, if not it will be long.

I think its still better to run through the list only once (feature) than twice (feature+hasfeature).

**Pull request merged**